src/coordination/coordinator_state_machine.cpp

```diff
@@ -12,14 +12,120 @@
+  nlohmann::json cluster_state_json;
+  memgraph::coordination::to_json(cluster_state_json, snapshot_ctx.cluster_state_);
+
+  nlohmann::json last_config_json;
+  memgraph::coordination::to_json(last_config_json, *snapshot_ctx.snapshot_->get_last_config());
+
+  j = nlohmann::json{{kCoordClusterState.data(), cluster_state_json.dump()},
+                     {kLastLogTerm.data(), snapshot_ctx.snapshot_->get_last_log_term()},
+                     {kLastLogIdx.data(), snapshot_ctx.snapshot_->get_last_log_idx()},
+                     {kSize.data(), snapshot_ctx.snapshot_->size()},
+                     {kLastConfig.data(), last_config_json.dump()},
+                     {kType.data(), static_cast<int>((snapshot_ctx.snapshot_->get_type()))}};
+}
+
+CoordinatorStateMachine::CoordinatorStateMachine(LoggerWrapper logger,
+                                                 std::optional<LogStoreDurability> log_store_durability)
+    : logger_(logger) {
+  if (log_store_durability.has_value()) {
+    durability_ = log_store_durability->durability_store_;
+  }
+  if (!durability_) {
+    logger_.Log(nuraft_log_level::WARNING, "Coordinator state machine stores snapshots only in memory from now on.");
+    return;
+  }
+
+  logger_.Log(nuraft_log_level::INFO, "Restoring coordinator state machine with durability.");
+
+  bool const successful_migration =
+      HandleMigration(log_store_durability->stored_log_store_version_, log_store_durability->active_log_store_version_);
+
+  MG_ASSERT(successful_migration, "Couldn't handle migration of log store version.");
+}
+
+bool CoordinatorStateMachine::HandleMigration(LogStoreVersion stored_version, LogStoreVersion active_version) {
+  if (stored_version == active_version && stored_version == LogStoreVersion::kV1) {
+    for (auto kv_store_snapshot_it = durability_->begin(std::string{kSnapshotIdPrefix});
+         kv_store_snapshot_it != durability_->end(std::string{kSnapshotIdPrefix}); ++kv_store_snapshot_it) {
+      auto const &[snapshot_key_id, snapshot_key_value] = *kv_store_snapshot_it;
+      try {
+        auto parsed_snapshot_id =
+            std::stoul(std::regex_replace(snapshot_key_id, std::regex{kSnapshotIdPrefix.data()}, ""));
+        last_committed_idx_ = std::max(last_committed_idx_.load(), parsed_snapshot_id);
+
+        auto snapshot_ctx_str = durability_->Get(snapshot_key_id);
+
+        if (!snapshot_ctx_str.has_value()) {
+          throw std::runtime_error("Failed to retrieve snapshot context from disk");
+        }
+
+        // NOLINTNEXTLINE (misc-const-correctness)
+        auto snapshot_ctx = cs_new<SnapshotCtx>();
+        from_json(nlohmann::json::parse(snapshot_ctx_str.value()), *snapshot_ctx);
+        snapshots_[parsed_snapshot_id] = snapshot_ctx;
+        MG_ASSERT(parsed_snapshot_id == snapshots_[parsed_snapshot_id]->snapshot_->get_last_log_idx(),
+                  "Parsed snapshot id {} does not match last log index {}", parsed_snapshot_id,
```

Comment:
If it fails it will crash completely, so the order is not that important in my opinion, but it makes sense to assert before you store to variable. I think it does, because otherwise we have really faulty logic, and we shouldn't continue with execution